include/deal.II/base/mpi.h

```diff
@@ -268,6 +272,71 @@ namespace Utilities
       MPI_Comm comm;
     };
 
+    /**
+     * An implementation of a critical section that guards an algorithm with MPI
+     * communication when executed several times.
+     *
+     * This critical section enforces that a section of code (between the
+     * constructor and destructor of this helper class) only starts executing
+     * after all MPI ranks in the same communicator finished execution of this
+     * section from an earlier time.
+     *
+     * The class internally executes a non-blocking barrier in the destructor
+     * for each critical section and waits for the completion of the barrier
+     * when constructed.
+     *
+     * This makes certain MPI algorithms correct that rely on receiving messages
+     * using MPI_ANY_SOURCE, where executing the same algorithm more than once
+     * in a row can confuse messages between the first and second execution.
+     *
+     * Example usage:
+     * @code
+     * {
+     *   static MPI_Request request = MPI_REQUEST_NULL;
+     *   CriticalSection cs(comm, request);
```

Comment:
If you do it this way, you have no way to clean up the last request because you lose track of the request object. 

The canonical design pattern would be if you had the `P2PCriticalSectionGuard` class have the request object as a member, and then have a scoped lock class that accesses the guard:
```
  static P2PCriticalSectionGuard critical_section_guard;
  P2PCriticalSectionGuard::ScopedLock lock(critical_section_guard);
```
This way, the `P3PCriticalSectionGuard` can register a call-back with the `MPI_InitFinalize` class that calls `MPI_Request_free` or `MPI_Wait` on the request at the end of the program.